### DISCUSSION TITLE
Avenir de Gipsy

### DISCUSSION THREAD

**Discussion by VForiel** [Gunivers/Gipsy#10] [General]
> Participez à la discussion via les commentaires ci-dessous. Ce premier message sert de récapitulatif et sera édité au fur et à mesure de l'avancée de la discussion.

### Quels sont les besoins ?

1. Une série de fonctionnalités permettant aux discords d'Altearn et ses communautés de fonctionner correctement.
2. Un bot modulaire dont l'écriture de plugin est simplifié au maximum. Ce bot pourra être aussi bien utilisé sur Altearn et ses communautés que hébergé par des personnes externe et/ou fournit à des clients.

Gipsy est un projet visant à répondre au premier besoin.
Pour le second besoin, nous utiliserons Zbot pour les fonctionnalités pouvant être intéressantes pour ce bot, et Gipsy pour les fonctionnalités spécifiques à nos communautés.

### Quels sont les problèmes actuels de Gipsy ?

L'écriture des plugins est actuellement trop lourd.
- Pour mettre en place des variables de configuration, plusieurs fichier doivent être édité et plusieurs dizaines de lignes doivent être ajoutés.
- Pour sauvegarder et récupérer des données persistantes, l'utilisation du SQL, la déclaration d'une base de donnée etc. sont nécessaires.
- Si une commande contient une erreur, c'est au plugin d'indiquer le problème.

### Idées d'amélioration

- Le bot va être réécrit "from scratch" avec un focus sur l'accessibilité, notamment avec l'utilisation de Sqlalchemy à la place d'instructions SQL brute
- Les plugins seront entièrement séparés du repo du cœur pour éviter, comme sur l'ancienne version de développer le cœur en fonction des plugin
- Faire un [suivie d'architecture](https://blog.stack-labs.com/code/adr-to-remember-past-archiectural-decisions/) qui sera particulièrement utile pour des nouveaux arrivants et, dans le cas d'une rewrite, pour éviter les choix peu judicieux
- Pour éviter de rendre l'enregistrement de commandes compliqué via le plugin : mettre une interface qui gérera la désactivation de commandes en fonction des paramètres et de gérer l'aide et autres informations (par exemple gipsy.command() avec le support des traductions de l'aide...)
- Si le but du projet est de proposer une solution à des "clients", il me paraîtrait judicieux de mettre en place un système de mise à jour de schéma de base de données pour éviter la perte éventuelle de données

### Proposition de réécriture

- Proposition de Leirof : #12

**Comment by ZRunner** [comment]
> Pour le second besoin, nous utiliserons Zbot pour les fonctionnalités pouvant être intéressantes pour ce bot,

Juste pour préciser, qu'entends-tu par ce point ?

Pour clarifier les choses sur Zbot, il n'est actuellement absolument pas capable d'être hébergé par quelqu'un d'autre. Entre les fichiers de sécurité cachés, les IDs hardcodées et la base de donnée privée, c'est tout bonnement impossible. De même pour le principe de plugins, qui bien qu'étant en cours d'amélioration depuis six mois, est encore loin d'être vraiment modulable (le premier exemple est celui de la base de donnée encore une fois). Il gère à peu près la désactivation des plugins, et encore, mais c'est loin d'être un vrai système de plugins téléchargeables et prêt à l'emploi.

Il faudrait un travail incroyablement massif sur Zbot avant de le rendre modulable et utilisable par tous, parce que ce n'est absolument pas la direction que je lui ai donné, ce n'est tout simplement pas dans mes plans.

Je pense que l'ajout de fonctionnalités sur Zbot qui pourraient intéresser un grand nombre de serveurs en plus de Gunivers/Altearn est tout à fait envisageable, c'est déjà ce que j'ai fait en solo pour le système de tickets par exemple. Mais si par "pour le second besoin nous utiliserons Zbot" tu entends le rendre téléchargeable et modulable, ça risque d'être compliqué.

**Reply by ascpial** [reply]
Je penses qu'il a inversé les besoins. Le bot modulable proposé aux clients sera Gipsy, alors que les fonctionnalités viendront de ZBot

**Reply by VForiel** [reply]
En effet ça n'est pas ce que je voulais dire.

Le fait est qu'on aura 2 bots à maintenir : Zbot et Gipsy.

Quitte à maintenir les deux bots, autant éviter de devoir maintenir deux codes ayant des objectifs communs. Ainsi, toutes les fonctionnalités "grand public" seront à mettre coté Zbot (si toutefois tu les trouve intéressantes @ZRunner), et toutes les autres (fonctionnalités propres à Gunivers et celles refusée sur Zbot) viendront sur Gipsy.

Gipsy viendra donc en complément de Zbot du point de vue de Gunivers, et sera un bot modulaire qu'on fournira en self-hostable et/ou à des clients voulant des fonctionnalités spécifiques. Zbot quand à lui sera un bot public non self-hostable sur lequel on participera à développer ses fonctionnalités.

**Comment by ascpial** [comment]
Il faut aussi ajouter comme problème dans gipsy le fait qu'il ne permet pas de gérer les fonctionnalités récentes de discord (notamment les commandes d'applications)
Il faut donc que la nouvelle architecture simplifie aussi tout ce qui touche à des enregistrements de commandes et autre.

**Reply by VForiel** [reply]
Il faudrait discuter de ce qui est possible à faire de ce coté là, pck ça me parait très dépendant de la lib, et j'imagine pas vraiment faire une surcouche à la lib

**Comment by ascpial** [comment]
Pour ce qui est des propositions, je résume ce qui a été décidé sur le Discord :
* le bot va être réécrit "from scratch" avec un focus sur l'accessibilité, notamment avec l'utilisation de Sqlalchemy à la place d'instructions SQL brute
* les plugins seront entièrement séparés du repo du cœur pour éviter, comme sur l'ancienne version de développer le cœur en fonction des plugin
* Oromis a proposé de faire un [suivie d'architecture](https://blog.stack-labs.com/code/adr-to-remember-past-archiectural-decisions/) qui sera particulièrement utile pour des nouveaux arrivants et, dans le cas d'une rewrite, pour éviter les choix peu judicieux
Mes propositions :
* pour éviter de rendre l'enregistrement de commandes compliqué via le plugin, je propose de mettre une interface qui gérera la désactivation de commandes en fonction des paramètres et de gérer l'aide et autres informations (par exemple `gipsy.command()` avec le support des traductions de l'aide...)
* si le but du projet est de proposer une solution à des "clients", il me paraîtrait judicieux de mettre en place un système de mise à jour de schéma de base de données pour éviter la perte éventuelle de données

**Comment by VForiel** [comment]
Message principal mis à jour

**Comment by Aeris1One** [comment]
Sur un autre point qui me semble important, la licence de Gipsy c'est actuellement du "Nothing here..."...
J'avais cru voir une licence Creative Commons pointer le bout de son nez il y a quelques temps, mais le problème de Creative Commons c'est que c'est absolument pas adapté à du logiciel. La fondation Creative Commons elle-même conseille d'utiliser une licence de copyleft prévue pour le logiciel, et elle-même utilise la licence MIT et la GNU GPL pour ses morceaux de code.

En gros, la MIT c'est "faites ce que vous voulez tant que vous créditez les auteurs originaux"
La GNU GPL c'est "faites ce que vous voulez tant que vous laissez le code sous licence GPL (et donc open-source)"
C'est les deux plus grosses licences, il y en a d'autres cependant.

Je pense qu'il serait bon de passer sous licence AGPL, il s'agit d'une licence GPL étendue avec, en gros "dès que vous utilisez le code modifié, vous devez le partager" là où la GPL dit seulement "vous devez partager le code source si vous repartagez".
Dans le cas d'un bot Discord la GPL simple n'obligerait à publier que si le bot modifié est proposé en self-host au public, là où l'AGPL demande à tous ceux qui utilisent le code dans un bot public de publier le code. Par "publier le code" la GPL/AGPL n'entends pas forcément "mettre sur Github" hein, mais juste bêtement "donner le code si on vous demande".
